### DISCUSSION TITLE
Debugging superenv

### DISCUSSION THREAD

**Discussion by cglong** [Homebrew/discussions#2533] [Tap maintenance and Homebrew development] [answered]
I'm trying to debug some build failures on an unsupported system and believe the root causes are in `superenv`. Are there any troubleshooting guides or recommended steps to follow to help me get started?

**Comment by carlocab** [comment, marked as answer]
Not really. What kind of issues are you seeing?

**Reply by cglong** [reply]
I've been investigating these two. Neither repros when built using the same steps outside Homebrew:

* rust-lang/rust#88068: `superenv` prepends Xcode include directories before libc++ ones, causing build failures ([more details from LLVM](https://bugs.llvm.org/show_bug.cgi?id=52575#c1))
* kubernetes/kubernetes#106572: `superenv` seems to be interfering with cgo, causing a linking error

**Reply by carlocab** [reply]
I'm not sure what to make of your Go error, but I don't think superenv adds Xcode include directories to `-isystem`. Here's where the macOS specific things are set: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/os/mac/extend/ENV/super.rb

What's the output of `brew config` and `brew doctor`?

**Reply by cglong** [reply]
> I don't think superenv adds Xcode include directories to `-isystem`.

Oh interesting, I wonder if there's something else about the Homebrew environment that's causing this (e.g. maybe the Rust CMake logic detects something and adds them?).

> What's the output of `brew config` and `brew doctor`?

* `brew config`: [config.log](https://github.com/Homebrew/discussions/files/7600050/config.log)
* `brew doctor`: [doctor.log](https://github.com/Homebrew/discussions/files/7600051/doctor.log) 

I think the Kubernetes mystery is solved 🙂 Still weird to me that it's attempting to get linked in Homebrew but not normally. Either way, I don't think I need `libxar`, so thank you @carlocab for your help!

Does anything stick out to you re: `-isystem`? Either way, I'll start looking through the `superenv` logic to see if I can find anything 🙂

**Reply by carlocab** [reply]
Nothing in particular. You should try fixing up your `brew doctor` warnings, though. Do you still have anything at `~/Library/Logs/Homebrew/rust`? I just remembered there should be more complete logs there about what superenv did and didn't add to compiler invocations.

If there isn't anything anymore, you can try producing your error again then do `brew gist-logs rust`.

**Reply by cglong** [reply]
> Do you still have anything at `~/Library/Logs/Homebrew/rust`? I just remembered there should be more complete logs there about what superenv did and didn't add to compiler invocations.

Oh nice! Yeah it looks like you're right @carlocab; I wasn't able to find any instances of superenv adding anything to do with Xcode. I also see that the CMake output is stored in there. That'll be super useful for debugging this 😄 Complete artifacts here: [rust.zip](https://github.com/Homebrew/discussions/files/7604943/rust.zip).

> You should try fixing up your `brew doctor` warnings, though.

Yeah there's definitely some things I should take care of there. The vast majority though is actually some other bug in Homebrew that appeared a few months ago. Every time `brew update` runs, it dirties its own tree. I usually have to do this:

```console
brew update && brew upgrade
pushd /usr/local/Homebrew && git reset --hard origin/master & popd
brew update && brew upgrade
```

I haven't had time to sit down and analyze Homebrew's `.gitignore` structure to figure out what's causing this 😅 Fortunately it doesn't really seem to cause any big problems.

**Reply by Bo98** [reply]
Looks like CMake is picking up the 10.14 SDK.

Does removing this line fix the issue? https://github.com/Homebrew/homebrew-core/blob/c334c367a6e4aa820fda8f56b153f630db7d22cb/Formula/rust.rb#L69

(not sure if it's still needed, but it is likely causing CMake to search for SDKs)

**Reply by cglong** [reply]
Thank you @Bo98!! It now builds successfully 🎉 It looks like that line was originally added in Homebrew/homebrew-core#20966 with description:

> set MACOSX_DEPLOYMENT_TARGET or compiler_builtins fails to build

Does this mean anything to you or @jyn514? If not, I may just open a PR so we can review the CI results 🙂

**Reply by Bo98** [reply]
It's probably a years old issue that doesn't apply anymore. If it fixes things for 10.13 I'm fine with a PR for it.

If something is truly needed for even older macOS (though Rust doesn't actually compile on 10.11 for different reasons anyway), then it will likely come in a different form of fix as `MACOSX_DEPLOYMENT_TARGET` is making CMake do behaviour we definitely do not want.

**Reply by jyn514** [reply]
Yeah, that PR title doesn't mean anything to me. I haven't actually ever used Mac.

**Reply by carlocab** [reply]
I think @EricFromCanada removed [that line](https://github.com/Homebrew/homebrew-core/blob/c334c367a6e4aa820fda8f56b153f630db7d22cb/Formula/rust.rb#L69) from `rust.rb` a while back precisely to fix building on older macOS. Not sure why it's still there. Maybe the change was cherry-picked into another PR and misplaced in the process.

**Edit:** Ah, no, that was `SDKROOT`. https://github.com/Homebrew/homebrew-core/pull/76780

**Reply by EricFromCanada** [reply]
The presence of `ENV["SDKROOT"] = MacOS.sdk_path` in several formulae causes macOS Sierra (and earlier?) to fail with "C compiler cannot create executables", unless `xcode-select -p` is set to the CLT instead of Xcode. It was usually added long ago to fix some other error that no longer occurs on newer OSes and so can usually be removed. But, the instances with the comment about fixing "ld: file not found: /usr/lib/system/libsystem_darwin.dylib" still apply to 10.12, because [the 10.13 SDK references a library](https://developer.apple.com/forums/thread/87829) [not included with 10.12](https://github.com/Homebrew/homebrew-core/pull/20872#issuecomment-345651099) which gets picked up whenever a build script calls `xml2-config --libs`. Essentially, on 10.12, any formulae with that SDKROOT fix can only be compiled if the CLT is active.

**Reply by Bo98** [reply]
> because [the 10.13 SDK references a library](https://developer.apple.com/forums/thread/87829) [not included with 10.12](https://github.com/Homebrew/homebrew-core/pull/20872#issuecomment-345651099) which gets picked up whenever a build script calls `xml2-config --libs`.

We could use shims to stop this. In fact I added one for `xcrun --show-sdk-path` a while back, but it currently only applies on Mojave and later.

**Reply by cglong** [reply]
Just opened Homebrew/homebrew-core#89988 to fix this!! Thank you everyone for all your help investigating this; I really appreciate it 😊

> Ah, no, that was `SDKROOT`.

That's great news actually; that was the other change included in Homebrew/homebrew-core#20966. The fact that it was successfully removed recently gives some evidence that this change is no longer needed either!
